## Pandas - Data Filtering

### 1. Introduction

Data Filtering is the process of drilling down a dataset to extract a subset that is relevant for analysis and decision-making. In today’s world of Big Data, we are usually confronted with datasets containing millions of rows and hundreds of columns. Understanding the filtering process in whatever tool is being used is a critical step for data analysis.

The Pandas package in Python is one of the best tools for manipulating data, and it includes a number of ways for filtering data. Today, I will explain and demonstrate seven (7) of these approaches for single and multiple condition filtering across rows and columns. These methods are:

* Column Names

* Query

* Nlargest/Nsmallest

* Operators (Comparison and Logical)

* Loc/Iloc 

* Isin

* String methods.

### 2. Importing Libraries and Loading Dateset

The core data manipulation package pandas and numpy where used for analysis. While the dataset used for his project is the Mexican Toy Store sales file, which was sourced from the @Maven_Analytics data playground.

In [1]:
#import libraries
import numpy as np
import pandas as pd

In [2]:
#loading dataset
toy_sales = pd.read_csv('C:\\Users\\User\\Documents\\maven_toys.csv')

print(toy_sales.shape)
toy_sales.head()

(65535, 8)


Date      Product_Name Product_Category  Units  Product_Price  \
0  1/1/2017  Chutes & Ladders            Games      1          12.99   
1  1/1/2017  Chutes & Ladders            Games      1          12.99   
2  1/1/2017  Chutes & Ladders            Games      1          12.99   
3  1/1/2017  Chutes & Ladders            Games      1          12.99   
4  1/1/2017  Chutes & Ladders            Games      1          12.99   

                    Store_Name      Store_City Store_Location  
0  Maven Toys Aguascalientes 1  Aguascalientes       Downtown  
1  Maven Toys Aguascalientes 1  Aguascalientes       Downtown  
2  Maven Toys Aguascalientes 1  Aguascalientes       Downtown  
3  Maven Toys Aguascalientes 1  Aguascalientes       Downtown  
4  Maven Toys Aguascalientes 1  Aguascalientes       Downtown

### 3 Data Exploration and Cleaning

In [3]:
#checking for null values and column datatypes
toy_sales.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Date              65535 non-null  object 
 1   Product_Name      65535 non-null  object 
 2   Product_Category  65535 non-null  object 
 3   Units             65535 non-null  int64  
 4   Product_Price     65535 non-null  float64
 5   Store_Name        65535 non-null  object 
 6   Store_City        65535 non-null  object 
 7   Store_Location    65535 non-null  object 
dtypes: float64(1), int64(1), object(6)
memory usage: 4.0+ MB


The dataset has 10 columns, 6 object types and 3 numeric (3 floats and 1 int) and 1 datetime column which will be converted in the next code block.

In [4]:
#Converting date column to datetime format
toy_sales['Date'] = pd.to_datetime(toy_sales['Date'], format='%m/%d/%Y')

In [5]:
toy_sales.dtypes

Date                datetime64[ns]
Product_Name                object
Product_Category            object
Units                        int64
Product_Price              float64
Store_Name                  object
Store_City                  object
Store_Location              object
dtype: object

In [6]:
#summarizing key statistical measures for numerical variables
toy_sales.describe()

Units  Product_Price
count  65535.000000   65535.000000
mean       1.288487      14.589619
std        0.766532       8.148112
min        1.000000       2.990000
25%        1.000000       8.990000
50%        1.000000      14.990000
75%        1.000000      15.990000
max       30.000000      39.990000

In [7]:
#checking cardinality of categorical variablies
cat_vars = toy_sales.select_dtypes(include='object')
nunique = cat_vars.nunique()

#creating a df
unique_values = pd.DataFrame({'Column': nunique.index, 'Unique Values': nunique.values})
print(unique_values)

             Column  Unique Values
0      Product_Name             25
1  Product_Category              5
2        Store_Name             50
3        Store_City             29
4    Store_Location              4


In [8]:
#value counts of items in specific columns
category = toy_sales[['Product_Category']].value_counts().rename_axis('Category').to_frame('No_Produts').reset_index()
print(category)

location = toy_sales[['Store_Location']].value_counts().rename_axis('Type').to_frame('No_Locations').reset_index()
print(location)

cities = toy_sales[['Store_City']].value_counts().rename_axis('City').to_frame('No_of_Records').reset_index()
print(cities)

            Category  No_Produts
0               Toys       19481
1              Games       14386
2        Electronics       13094
3  Sports & Outdoors        9740
4       Art & Crafts        8834
          Type  No_Locations
0     Downtown         38374
1   Commercial         14558
2  Residential          7748
3      Airport          4855
                City  No_of_Records
0   Cuidad de Mexico           7060
1        Guadalajara           5971
2          Monterrey           5409
3         Guanajuato           3938
4             Puebla           3810
5         Hermosillo           3772
6           Mexicali           3078
7             Toluca           2920
8          Chihuahua           2672
9           Saltillo           2651
10          Campeche           2637
11            Xalapa           2380
12    Aguascalientes           1469
13   San Luis Potosi           1453
14      Villahermosa           1296
15           Morelia           1276
16   Ciudad Victoria           1272
17       

In [9]:
#stripping Maven Toys from Store_Name column
toy_sales['Store_Name'] = toy_sales['Store_Name'].str.replace('Maven Toys', '')

In [10]:
#creating new dfs for product sales and store stores
toy_sales['Revenue'] = toy_sales['Product_Price'] * toy_sales['Units']

In [11]:
#product sales
product_sales = toy_sales.groupby('Product_Name', as_index = False).agg( Quantity = ('Product_Name', 'count' ),
                                                                        Sales = ('Revenue', 'sum'))
                                                            
product_sales.head(10)                                                             

Product_Name  Quantity      Sales
0     Action Figure      6327  118198.08
1    Animal Figures      3126   49063.23
2   Barrel O' Slime        94     375.06
3  Chutes & Ladders       207    2753.88
4  Classic Dominoes       489    4885.11
5         Colorbuds     11511  271588.82
6          Dart Gun      3186   63160.50
7     Deck Of Cards      7056   62406.72
8          Dino Egg      3422   47388.88
9  Dinosaur Figures      1831   33157.88

In [12]:
#store sales
store_sales = toy_sales.groupby('Store_Name', as_index = False).agg( Quantity = ('Product_Name', 'count' ),
                                                                        Sales = ('Revenue', 'sum'))
                                                            
store_sales.head(10)

Store_Name  Quantity     Sales
0     Aguascalientes 1      1469  26240.37
1           Campeche 1      1402  23564.27
2           Campeche 2      1235  21230.21
3           Chetumal 1      1189  22391.15
4          Chihuahua 1      1256  23420.62
5          Chihuahua 2      1416  25248.40
6       Chilpancingo 1      1021  16753.92
7    Ciudad Victoria 1      1272  25696.72
8   Ciudad de Mexico 1      2352  46942.87
9   Ciudad de Mexico 2      2012  46686.49

### 4. Filtering Methods

Five filtering that will be demonstrated are; Query , Nlargest/Nsmallest, Loc/Iloc, Isin and String methods

**4.1 - df[['column1', 'column2']]**: One of the quickests ways of filtering a dataframe is by selecting only specific columns needed for analysis at that point in time.

Note: Double Braces [[        ]] are used while subsetting dataframes using column names.The inner square brackets define a Python list with column names, whereas the outer brackets are used to select the data from the DataFrame.

In [13]:
specific_df = toy_sales[['Product_Category','Store_City','Revenue']]

print(specific_df.shape)
specific_df.head()

(65535, 3)


Product_Category      Store_City  Revenue
0            Games  Aguascalientes    12.99
1            Games  Aguascalientes    12.99
2            Games  Aguascalientes    12.99
3            Games  Aguascalientes    12.99
4            Games  Aguascalientes    12.99

The new DataFrame contains only 3 columns needed for a specific analysis.

**4.2 - df.query( )**: This method is used to filter rows based on single or multiple conditions. It is my go-to filtering method because it offers flexibility while writing conditions and it also works with both string and numeric columns.

In [14]:
#filtering sales details of stores located at the airport - single condition
airport1 = toy_sales.query("Store_Location == 'Airport'").reset_index(drop = True)

print(airport1.shape)
airport1.head()

(4855, 9)


Date      Product_Name Product_Category  Units  Product_Price  \
0 2017-01-28  Chutes & Ladders            Games      1          12.99   
1 2017-01-28  Chutes & Ladders            Games      1          12.99   
2 2017-01-28  Chutes & Ladders            Games      1          12.99   
3 2017-01-28  Chutes & Ladders            Games      1          12.99   
4 2017-02-06  Chutes & Ladders            Games      1          12.99   

     Store_Name Store_City Store_Location  Revenue  
0   Monterrey 3  Monterrey        Airport    12.99  
1   Monterrey 3  Monterrey        Airport    12.99  
2   Monterrey 3  Monterrey        Airport    12.99  
3   Monterrey 3  Monterrey        Airport    12.99  
4   Monterrey 3  Monterrey        Airport    12.99

The new dataframe contains sales from only airport location. Sales record reduces from 65.5k
items to 4.85k items

In [15]:
#filtering sales for items with product price > 19.99 at airport locations - multiple conditions
airport2 = toy_sales.query("Store_Location == 'Airport' & Product_Price > 19.99").reset_index(drop = True)

print(airport2.shape)
airport2.head()

(369, 9)


Date Product_Name Product_Category  Units  Product_Price  \
0 2017-01-06  Lego Bricks             Toys      1          39.99   
1 2017-01-06  Lego Bricks             Toys      1          39.99   
2 2017-01-06  Lego Bricks             Toys      1          39.99   
3 2017-01-29  Lego Bricks             Toys      1          39.99   
4 2017-01-29  Lego Bricks             Toys      1          39.99   

     Store_Name Store_City Store_Location  Revenue  
0   Monterrey 3  Monterrey        Airport    39.99  
1   Monterrey 3  Monterrey        Airport    39.99  
2   Monterrey 3  Monterrey        Airport    39.99  
3   Monterrey 3  Monterrey        Airport    39.99  
4   Monterrey 3  Monterrey        Airport    39.99

Airport location sales was further filtered to get expensive items sold at these location. The new dataframe now has 369 records.

In [16]:
#instead of selecting, value(s) can also be excluded using the != (not equal to) symbol
other_products = toy_sales.query("Product_Name != ['Action Figure', 'Deck of Cards', 'Dino Egg','Colorbuds']")

print(other_products.shape)
other_products.head()                       

(44275, 9)


Date      Product_Name Product_Category  Units  Product_Price  \
0 2017-01-01  Chutes & Ladders            Games      1          12.99   
1 2017-01-01  Chutes & Ladders            Games      1          12.99   
2 2017-01-01  Chutes & Ladders            Games      1          12.99   
3 2017-01-01  Chutes & Ladders            Games      1          12.99   
4 2017-01-01  Chutes & Ladders            Games      1          12.99   

          Store_Name      Store_City Store_Location  Revenue  
0   Aguascalientes 1  Aguascalientes       Downtown    12.99  
1   Aguascalientes 1  Aguascalientes       Downtown    12.99  
2   Aguascalientes 1  Aguascalientes       Downtown    12.99  
3   Aguascalientes 1  Aguascalientes       Downtown    12.99  
4   Aguascalientes 1  Aguascalientes       Downtown    12.99

**4.3 - nlargest() and nsmallest():** This function returns the first n rows in either descending (nlargest) or ascending (nsmallest) order. It is, in my opinion, one of the simplest methods to implement. The method comes in handy when running rapid filtering to determine the top or bottom categories in a dataset. This approach, however, only works with numerical columns.

In [17]:
#filtering for the top 5 selling stores
top_stores = store_sales.nlargest(5, 'Sales').reset_index(drop = True)
top_stores

Store_Name  Quantity     Sales
0   Ciudad de Mexico 1      2352  46942.87
1   Ciudad de Mexico 2      2012  46686.49
2        Guadalajara 4      1827  36965.63
3             Toluca 1      1963  35845.47
4           Mexicali 1      1658  31678.10

In [18]:
#filtering for lowest quantity sold by products
lowest_products = product_sales.nsmallest(3, 'Quantity').reset_index(drop = True)
lowest_products

Product_Name  Quantity   Sales
0      Uno Card Game        46  367.54
1    Barrel O' Slime        94  375.06
2  Hot Wheels 5-Pack       117  790.68

**4.4- Operators (Comparison and Logical)**: Dataframes can also be filtered using Python Operators. Operators are special symbols that perform operations on variables and values.

Comparison operators like (>,<, ==, !=, >=, <=) are used to filter rows using column values while Logical operators like (&, or) are used to introduce multiple condition filtering.

In [19]:
#filtering column using comparison operator
expensive_toys = toy_sales[toy_sales.Product_Price > 20].reset_index(drop = True)

print(expensive_toys.shape)
expensive_toys.head()

(5708, 9)


Date Product_Name Product_Category  Units  Product_Price  \
0 2017-01-17  Lego Bricks             Toys      1          39.99   
1 2017-01-17  Lego Bricks             Toys      1          39.99   
2 2017-01-17  Lego Bricks             Toys      1          39.99   
3 2017-01-17  Lego Bricks             Toys      1          39.99   
4 2017-01-17  Lego Bricks             Toys      1          39.99   

          Store_Name      Store_City Store_Location  Revenue  
0   Aguascalientes 1  Aguascalientes       Downtown    39.99  
1   Aguascalientes 1  Aguascalientes       Downtown    39.99  
2   Aguascalientes 1  Aguascalientes       Downtown    39.99  
3   Aguascalientes 1  Aguascalientes       Downtown    39.99  
4   Aguascalientes 1  Aguascalientes       Downtown    39.99

The new dataframe contains only prodcuts with price > 20. Sales record reduces from 65.5k items to 5.7k items

In [20]:
expensive_electronics = toy_sales[(toy_sales.Product_Price > 20) & (toy_sales.Product_Category
                                                              == 'Electronics' )].reset_index(drop = True)

print(expensive_electronics.shape)
expensive_electronics.head()

(1583, 9)


Date Product_Name Product_Category  Units  Product_Price  \
0 2017-01-17    Toy Robot      Electronics      1          25.99   
1 2017-01-17    Toy Robot      Electronics      1          25.99   
2 2017-01-17    Toy Robot      Electronics      1          25.99   
3 2017-01-17    Toy Robot      Electronics      1          25.99   
4 2017-01-17    Toy Robot      Electronics      1          25.99   

          Store_Name      Store_City Store_Location  Revenue  
0   Aguascalientes 1  Aguascalientes       Downtown    25.99  
1   Aguascalientes 1  Aguascalientes       Downtown    25.99  
2   Aguascalientes 1  Aguascalientes       Downtown    25.99  
3   Aguascalientes 1  Aguascalientes       Downtown    25.99  
4   Aguascalientes 1  Aguascalientes       Downtown    25.99

After applying multiple condition filtering using logical operators, the new datarame now contains 1.58k records.

**4.5 - Loc/Iloc Method**: This filtering technique is best suited for extracting specific rows from a DataFrame. The twin methods create a subset of a DataFrame using label-based or integer-based indexing, respectively.

Loc selects rows and columns with specific lablels

Iloc selects rows and columns with specific integer values following a 0-based integer position system.

In [21]:
#filtering using .loc method to select specific rows and columns
loc_df = toy_sales.loc[100:499, ['Date', 'Product_Name','Units','Store_City']]

print(loc_df.shape)
loc_df.head()

(400, 4)


Date   Product_Name  Units      Store_City
100 2017-02-13  Action Figure      1  Aguascalientes
101 2017-02-16  Action Figure      1  Aguascalientes
102 2017-02-16  Action Figure      1  Aguascalientes
103 2017-02-16  Action Figure      1  Aguascalientes
104 2017-02-16  Action Figure      1  Aguascalientes

In [22]:
#filtering using .iloc method to select specific rows and columns
# since iloc uses 0-based integers,row 500 and the 5th column (Product_Price) are not included

iloc_df = toy_sales.iloc[100:500, 0:4]

print(iloc_df.shape)
iloc_df

(400, 4)


Date   Product_Name   Product_Category  Units
100 2017-02-13  Action Figure               Toys      1
101 2017-02-16  Action Figure               Toys      1
102 2017-02-16  Action Figure               Toys      1
103 2017-02-16  Action Figure               Toys      1
104 2017-02-16  Action Figure               Toys      1
..         ...            ...                ...    ...
495 2017-01-08   Splash Balls  Sports & Outdoors      1
496 2017-01-08   Splash Balls  Sports & Outdoors      1
497 2017-01-08   Splash Balls  Sports & Outdoors      1
498 2017-01-14   Splash Balls  Sports & Outdoors      1
499 2017-01-14   Splash Balls  Sports & Outdoors      1

[400 rows x 4 columns]

**4.6 Isin Method()**:This method pulls data from a column based on particular values. It determines whether the specified column contains any of the values from the isin() function's list then returns the rows containing the list values.

In [23]:
#filtering a list using isin() function
selected_cities = toy_sales[toy_sales['Store_City'].isin(['Toluca','La Paz','Puebla'])].reset_index(drop = True)

print(selected_cities.shape)
selected_cities.head()

(7935, 9)


Date   Product_Name Product_Category  Units  Product_Price Store_Name  \
0 2017-01-01  Action Figure             Toys      1          15.99   Puebla 2   
1 2017-01-01  Action Figure             Toys      1          15.99   Puebla 2   
2 2017-01-01  Action Figure             Toys      1          15.99   Puebla 2   
3 2017-01-01  Action Figure             Toys      1          15.99   Puebla 2   
4 2017-01-01  Action Figure             Toys      1          15.99   Puebla 2   

  Store_City Store_Location  Revenue  
0     Puebla       Downtown    15.99  
1     Puebla       Downtown    15.99  
2     Puebla       Downtown    15.99  
3     Puebla       Downtown    15.99  
4     Puebla       Downtown    15.99

Using the isn() method, records from only stores in 3 cities Toluca, La Paz, and Puebla were selected.

In [24]:
#applying multiple condition filtering using isin() method
selected_cities2 = toy_sales[toy_sales['Store_City'].isin(['Toluca','La Paz','Puebla']) &
                            toy_sales['Product_Category'].isin(['Art & Crafts'])].reset_index(drop = True)

print(selected_cities2.shape)
selected_cities2.head()

(1474, 9)


Date     Product_Name Product_Category  Units  Product_Price  \
0 2017-01-02  PlayDoh Toolkit     Art & Crafts      1           4.99   
1 2017-01-02  PlayDoh Toolkit     Art & Crafts      1           4.99   
2 2017-01-02  PlayDoh Toolkit     Art & Crafts      1           4.99   
3 2017-01-02  PlayDoh Toolkit     Art & Crafts      1           4.99   
4 2017-01-02  PlayDoh Toolkit     Art & Crafts      1           4.99   

  Store_Name Store_City Store_Location  Revenue  
0   Puebla 2     Puebla       Downtown     4.99  
1   Puebla 2     Puebla       Downtown     4.99  
2   Puebla 2     Puebla       Downtown     4.99  
3   Puebla 2     Puebla       Downtown     4.99  
4   Puebla 2     Puebla       Downtown     4.99

Multiple filtering was applied to selected on Arts & Craft items sold into the 3 cities selected ealier

**4.7 String Methods**: Pandas is a very efficient textual data library. The str accessor's functions and methods give versatile ways to filter rows based on strings. 
These methods computes patterns to extract data from the dataframe by using methods such as str.contains() and str.startswith() to determine whether a string contains the needed pattern.

In [25]:
#filtlering dataframe for only residential and commericial sales using str.contains()
rescomm_sales = selected_cities = toy_sales[toy_sales['Store_Location'].str.contains
                                             ('Residential|Commerical')].reset_index(drop = True)

print(rescomm_sales.shape)
rescomm_sales.head()

(7748, 9)


Date      Product_Name Product_Category  Units  Product_Price  \
0 2017-01-02  Chutes & Ladders            Games      1          12.99   
1 2017-01-02  Chutes & Ladders            Games      1          12.99   
2 2017-01-02  Chutes & Ladders            Games      1          12.99   
3 2017-01-02  Chutes & Ladders            Games      1          12.99   
4 2017-01-02  Chutes & Ladders            Games      1          12.99   

            Store_Name        Store_City Store_Location  Revenue  
0   Ciudad de Mexico 3  Cuidad de Mexico    Residential    12.99  
1   Ciudad de Mexico 3  Cuidad de Mexico    Residential    12.99  
2   Ciudad de Mexico 3  Cuidad de Mexico    Residential    12.99  
3   Ciudad de Mexico 3  Cuidad de Mexico    Residential    12.99  
4   Ciudad de Mexico 3  Cuidad de Mexico    Residential    12.99

In [26]:
#filtlering dataframe for only sales from the city of Monterrey using str.startswith()
monterrey_sales = toy_sales[toy_sales['Store_City'].str.startswith('Monterrey')].reset_index(drop = True)

print(monterrey_sales.shape)
monterrey_sales.head()

(5409, 9)


Date      Product_Name Product_Category  Units  Product_Price  \
0 2017-01-01  Chutes & Ladders            Games      1          12.99   
1 2017-01-01  Chutes & Ladders            Games      1          12.99   
2 2017-01-01  Chutes & Ladders            Games      1          12.99   
3 2017-01-02     Action Figure             Toys      1          15.99   
4 2017-01-02     Action Figure             Toys      1          15.99   

     Store_Name Store_City Store_Location  Revenue  
0   Monterrey 2  Monterrey       Downtown    12.99  
1   Monterrey 2  Monterrey       Downtown    12.99  
2   Monterrey 2  Monterrey       Downtown    12.99  
3   Monterrey 2  Monterrey       Downtown    15.99  
4   Monterrey 2  Monterrey       Downtown    15.99

Filtering is a powerful tool for drilling into data to extraxt useful information per time, and the most frequently used methods and their best uses cases have been described in this notebook with examples.